### PR TITLE
updated pymatreader to version 0.0.12.

### PR DESCRIPTION
**Pull request by thht**
This version improves reading nested cell arrays in old version matlab files. thanks to @DiGyt

### CODE REVIEW

**Comment by thht** [comment]
ok, we definitely broke something here that did not show up in our tests. we will investigate and come back.

**Comment by larsoner** [comment]
We have a number of places we still triage based on type (old vs new), so it's also possible that you made things more consistent at the pymatreader level and we need to update our code.

**Comment by thht** [comment]
ok, looked at it. the problem seems to be that `scipy.io.loadmat` does not distinguish between a struct array (which is used by EEGlab to store the events) and cell arrays.

`_read_annotations_eeglab` expects them to be loaded in the more "original" way, i.e., for every field of the structured array, there should be one list with all the items. the v7.3 code does it correctly while the scipy based code treats it as if it was a cell array (i.e., makes it a list of dicts).

i have not found a way to fix this behavior using the output of `scipy.io.loadmat` as a cell array and a struct array look the same when it loads it.

i am going to try to work on the v7.3 code so a consistent behavior will be shown by the library. but anyhow, this would involve adapting the `_bunchify` function.

i have to do other stuff now and will update you next week.

File: mne/externals/pymatreader/pymatreader.py [hunk 1]
```diff
@@ -130,12 +133,15 @@ def _hdf5todict(hdf5_object, variable_names=None, ignore_fields=None):
...
-            return_dict[key] = _hdf5todict(hdf5_object[key], variable_names=None, ignore_fields=ignore_fields)
+            return_dict[key] = _hdf5todict(hdf5_object[key],
+                                           variable_names=None,
+                                           ignore_fields=ignore_fields)
 
         return return_dict
     elif isinstance(hdf5_object, h5py.Dataset):
         data = hdf5_object.value
-        if isinstance(data, numpy.ndarray) and data.dtype == numpy.dtype('object'):
+        if isinstance(data, numpy.ndarray) and data.dtype == numpy.dtype(
```
Comment (massich) [inline mne/externals/pymatreader/pymatreader.py]: why did we change this line?

if needed to change, I like better to break after `and \`

File: mne/externals/pymatreader/pymatreader.py [hunk 2]
```diff
@@ -161,7 +167,8 @@ def _assign_types(values):
         else:
             assigned_values = values
 
-        if isinstance(assigned_values, numpy.ndarray) and assigned_values.size == 1:
+        if isinstance(assigned_values,
```
Comment (massich) [inline mne/externals/pymatreader/pymatreader.py]: idem

File: mne/externals/pymatreader/pymatreader.py [hunk 3]
```diff
@@ -189,22 +196,20 @@ def _check_for_scipy_mat_struct(data):
...
 
-    if isinstance(data, numpy.ndarray) and data.dtype == numpy.dtype('object') and not isinstance(data, scipy.io.matlab.mio5.MatlabFunction):
-        as_list = data.tolist()
-        try:
-            for (element, list_element) in zip(numpy.nditer(data, flags=['refs_ok'], op_flags=['readwrite']), as_list):
-                if not (isinstance(list_element, numpy.ndarray) and not list_element.dtype == numpy.dtype('object')):
-                    element[...] = _check_for_scipy_mat_struct(list_element)
-        except TypeError:
-            pass
+    if isinstance(data, numpy.ndarray) and data.dtype == numpy.dtype(
```
Comment (massich) [inline mne/externals/pymatreader/pymatreader.py]:
```py
    if isinstance(data, numpy.ndarray) and \
       data.dtype == numpy.dtype('object') and \
       not isinstance(data, scipy.io.matlab.mio5.MatlabFunction):
```

**Comment by larsoner** [comment]
> but anyhow, this would involve adapting the _bunchify function.

Hopefully whatever we have to do reduces the number of places we need to triage old vs new style. No rush from our end since things currently work at least.

**Comment by thht** [comment]
> Hopefully whatever we have to do reduces the number of places we need to triage old vs new style.

yes, this is what we are trying to do. i actually have tests in place that check whether pymatreaders output is the same for old vs. new style files and it worked with files containing test data that i came up with as well as with data coming from fieldtrip. EEGlab seems to use some matlab types (especially said struct arrays) that are represented differently....

**Comment by thht** [comment]
circleci failes with this:

```
#!/bin/bash -eo pipefail
if [[ $(cat merge.txt) != "" ]]; then
  echo "Merging $(cat merge.txt)";
  git pull --ff-only origin "refs/pull/$(cat merge.txt)/merge";
fi

Merging 5429
fatal: Couldn't find remote ref refs/pull/5429/merge
Exited with code 1
```

what does that mean? do i do something wrong?

**Comment by larsoner** [comment]
It's because CircleCI is building on your fork rather than MNE's

**Comment by larsoner** [comment]
You can disable builds for your fork but really we should improve the circle config to handle this

**Comment by thht** [comment]
thanks @larsoner . i did some investigating in the failed tests. will work on those today or tomorrow...

**Comment by thht** [comment]
@larsoner: the tests have run through. only one of the travisci builds failed: https://travis-ci.org/mne-tools/mne-python/jobs/415347726 the error message indicates it might be a problem with the build itself. so maybe rerunning it would do the trick?

@massich: i made pymatreader follow python coding style rules. if i understood your review correctly, it was about this. the code is updated. may i ask you do review the current version?

**Comment by thht** [comment]
ok, retriggered myself and it's all green now.

**Comment by larsoner** [comment]
Thanks @thht

**Pull request merged**